spec/strings_file_validation_helper_spec.rb

```diff
@@ -5,6 +5,98 @@ module Fastlane
 
   module Helper
     class StringsFileValidationHelper
+      # context can be one of:
+      #   :root, :maybe_comment_start, :in_line_comment, :in_block_comment, :maybe_block_comment_end,
+      #   :in_quoted_key, :after_quoted_key_before_eq, :after_quoted_key_and_equal, :in_quoted_value, :after_quoted_value
+      State = Struct.new(:context, :buffer, :in_escaped_ctx, :found_key, keyword_init: true)
+
+      TRANSITIONS = {
+        root: {
+          /\s/ => :root,
+          '/' => :maybe_comment_start,
+          '"' => :in_quoted_key
+        },
+        maybe_comment_start: {
+          '/' => :in_line_comment,
+          /\*/ => :in_block_comment
+        },
+        in_line_comment: {
+          "\n" => :root,
+          /./ => :in_line_comment
+        },
+        in_block_comment: {
+          /\*/ => :maybe_block_comment_end,
+          /./m => :in_block_comment
+        },
+        maybe_block_comment_end: {
+          '/' => :root,
+          /./m => :in_block_comment
+        },
+        in_quoted_key: {
+          '"' => lambda do |state, _|
+            state.found_key = state.buffer.string.dup
+            state.buffer.string = ''
+            :after_quoted_key_before_eq
+          end,
+          /./ => lambda do |state, c|
+            state.buffer.write(c)
+            :in_quoted_key
+          end
+        },
+        after_quoted_key_before_eq: {
+          /\s/ => :after_quoted_key_before_eq,
+          '=' => :after_quoted_key_and_eq
+        },
+        after_quoted_key_and_eq: {
+          /\s/ => :after_quoted_key_and_eq,
+          '"' => :in_quoted_value
+        },
+        in_quoted_value: {
+          '"' => :after_quoted_value,
+          /./m => :in_quoted_value
+        },
+        after_quoted_value: {
+          /\s/ => :after_quoted_value,
+          ';' => :root
+        }
+      }.freeze
+
+      def self.find_duplicated_keys_using_state_machine(file:)
+        all_keys = {}
+        dup_keys = []
+
+        state = State.new(context: :root, buffer: StringIO.new, in_escaped_ctx: false, found_key: nil)
+
+        File.readlines(file).each_with_index do |line, line_no|
+          line.chars.each_with_index do |c, col_no|
+            # Handle escaped characters at a global level
+            if (!state.in_escaped_ctx && c == '\\') || state.in_escaped_ctx
+              state.buffer.write(c) if state.context == :in_quoted_key
+              state.in_escaped_ctx = !state.in_escaped_ctx
+              next
+            end
```

Comment:
I just realized that doing this at global level independently of the current context means the state machine accepts escaped chars in `:root` context (i.e. outside of any comment or quoted key or value), while in practice this is not valid.

We should:
 - either keep it at global level but add more conditions to the `if`, especially `&& [:in_quoted_key, :in_quoted_value].include?(state.context)`
 - or find a way to include this escape handling as part of the `TRANSITIONS` table — but going that route could prove a bit tricky and make it quite unreadable, because for example when in `:in_quoted_key` context, we'd have to act differently when matching `'"'` depending on if `in_escaped_ctx` is `true` (then just read the char in buffer) or `false` (do what the current `lambda` does). Otoh that would allow us to avoid having that escape processing being separate to the rest of the rules…

You might want to experiment with different solutions (handle escapes separately/globally vs at each state transition)  to find the one that is most readable.

Reply:
Dammit, "approved" the PR instead of just "commenting" apparently… but can't find a way to discard my accidentaly approval (probably because the PR is still Draft?). Hoping that re-requesting my own review will do it?

Reply:
> EDIT: Confirmed that I had to undrafting the PR to be able to dismiss my accidental approval (before drafting the PR again). 🤷